lib/charms/traefik_k8s/v0/ingress_per_unit.py

```diff
@@ -186,13 +431,102 @@ def is_failed(self, relation: Relation = None):
 
+    def get_request(self, relation: Relation):
+        """Get the IngressRequest for the given Relation."""
+        return IngressRequest(self, relation, self._fetch_ingress_data(relation))
+
+    def _fetch_ingress_data(self, relation: Relation, validate=False):
+        """Fetch and validate the databags.
+
+        For the provider side: the application databag.
+        For the requirer side: the unit databag.
+        """
+        this_unit = self.unit
+        this_app = self.app
+
+        if not relation.app or not relation.app.name:
+            # Handle edge case where remote app name can be missing.
+            return {relation.app: {}, this_app: {}, this_unit: {}}
+
+        ingress_data: dict = {}
+        # we start by looking at the provider's app databag
+        if this_unit.is_leader():
+            # only leaders can read their app's data
+            data = relation.data[this_app].get("data")
+            deserialized = {}
+            if data:
+                deserialized = _deserialize_data(data)
+                if validate:
+                    _validate_data(deserialized, INGRESS_PROVIDES_APP_SCHEMA)
+            ingress_data[this_app] = deserialized
+        else:
+            # non-leader units cannot read/write the app databag
+            ingress_data[this_app] = {}
+
+        # then look at the requirer's (thus remote) unit databags
+        remote_units = [
+            obj for obj in relation.data if isinstance(obj, Unit) and obj is not this_unit
+        ]
+
+        for remote_unit in remote_units:
+            remote_data = relation.data[remote_unit].get("data")
+            remote_deserialized = {}
+            if remote_data:
+                remote_deserialized = _deserialize_data(remote_data)
+                if validate:
+                    _validate_data(remote_deserialized, INGRESS_REQUIRES_UNIT_SCHEMA)
+            ingress_data[remote_unit] = remote_deserialized
+
+        return ingress_data
+
+    def publish_ingress_data(
+        self, relation: Relation, data: typing.Dict[typing.Union[Unit, Application], dict]
+    ):
+        """Publish ingress data to the relation databag."""
+        this_app = self.app
+        this_unit = self.unit
+
+        old_data = self._fetch_ingress_data(relation, validate=False)
+        for entity in data:
+
+            # validation step 1: check that we are not changing anything
+            # except the data buckets we have access to
+            if entity not in (this_app, this_unit):
+                # check that we're not attempting to push new data to remote
+                # units or apps
+                if data[entity] != old_data.get(entity):
+                    raise RelationPermissionError(relation, this_unit)
+                continue
+
+            # validation step 2: only leaders can write app data
+            if entity is this_app and not this_unit.is_leader():
+                raise RelationPermissionError(relation, this_unit)
```

Comment:
Can this case be covered by a regular leader guard at the beginning of the function?
And why raise in that case?